commands/generate.js

```diff
@@ -0,0 +1,102 @@
+'use strict';
+
+var url = require('url'),
+    path = require('path');
+
+var Q = require('q');
+
+var lib = require('manifoldjs-lib');
+
+var log = lib.log,
+    manifestTools = lib.manifestTools,
+    projectBuilder = lib.projectBuilder,
+    utils = lib.utils;
+
+var build = require('./package');
+
+function getW3cManifest(siteUrl, manifestLocation, callback) {
+  function resolveStartURL(err, manifestInfo) {
+    if (err) {
+      return callback(err, manifestInfo);
+    }
+
+    return manifestTools.validateAndNormalizeStartUrl(siteUrl, manifestInfo, callback);
+  }
+  
+  if (siteUrl) {
+    var parsedSiteUrl = url.parse(siteUrl);
+    if (!parsedSiteUrl.hostname) {
+      return callback(new Error('The site URL is not a valid URL.'));
+    }
+  }
+
+  if (manifestLocation) {
+    var parsedManifestUrl = url.parse(manifestLocation);
+    if (parsedManifestUrl && parsedManifestUrl.host) {
+      // download manifest from remote location
+      log.info('Downloading manifest from ' + manifestLocation + '...');
+      manifestTools.downloadManifestFromUrl(manifestLocation, resolveStartURL);
+    } else {
+      // read local manifest file
+      log.info('Reading manifest file ' + manifestLocation + '...');
+      manifestTools.getManifestFromFile(manifestLocation, resolveStartURL);
+    }
+  } else if (siteUrl) {    
+    // scan a site to retrieve its manifest
+    log.info('Scanning ' + siteUrl + ' for manifest...');
+    manifestTools.getManifestFromSite(siteUrl, resolveStartURL);
+  } else {
+    return callback(new Error('A site URL or manifest should be specified.'));
+  }
+}
+
+function generateApp(program) {
+  
+  var siteUrl = program.args[0];
+  var rootDir = program.directory ? path.resolve(program.directory) : process.cwd();
+  var platforms = program.platforms.split(/[\s,]+/);
+  
+  var deferred = Q.defer();  
+  getW3cManifest(siteUrl, program.manifest, function (err, manifestInfo) {
+    if (err) {
+      return deferred.reject(err);

```

Comment:
Hi @marcoscaceres

Thank you for taking the time to provide all this feedback. There was no time to include more changes for the current release, but we will take this into consideration for the next update.

This release involved significant re-architecture in a limited time and there are a number of things that had to be postponed. Regarding the use of standard promises, this is something that we considered and need to investigate.